File: src/eil/connectors/jira.py
"""Live Jira DC connector: cursor-based JQL incremental sync.

Same contract rule as Confluence: output is the issue dict the fixture path
uses, so eil.ingest.jira.normalize is the interface either way. Personal PAT
via EIL_JIRA_TOKEN in local mode.

Note: Jira DC descriptions/comments are wiki markup, not markdown — close
enough for FTS; a proper wiki->md pass can join the normalizer later without
changing this connector.
"""

from __future__ import annotations

import os
from collections.abc import Iterator
from typing import Any

from eil.connectors.auth import make_client
from eil.connectors.confluence import cql_ts

PAGE_SIZE = 50
FIELDS = "summary,description,status,issuetype,project,reporter,created,updated,comment"


class JiraClient:
    def __init__(self, base_url: str | None = None, token: str | None = None) -> None:
        self.base_url = (base_url or os.environ["EIL_JIRA_URL"]).rstrip("/")
        self.http = make_client("JIRA", self.base_url, token)

    def updated_since(self, cursor: str | None) -> Iterator[dict[str, Any]]:
        jql = "order by updated asc"
        if cursor:
            jql = f'updated >= "{cql_ts(cursor)}" order by updated asc'
        start = 0
        while True:
            resp = self.http.get(
                "/rest/api/2/search",
                params={"jql": jql, "fields": FIELDS, "maxResults": PAGE_SIZE, "startAt": start},
            )
            resp.raise_for_status()
            data = resp.json()
            issues = data.get("issues", [])
            for issue in issues:
                yield self.to_issue_dict(issue)
            start += len(issues)
            if start >= data.get("total", 0) or not issues:
                return

    def to_issue_dict(self, api_issue: dict[str, Any]) -> dict[str, Any]:
        f = api_issue["fields"]
        comments = [
            {"author": c.get("author", {}).get("displayName", "unknown"), "body": c.get("body", "")}
            for c in f.get("comment", {}).get("comments", [])
        ]
        return {
            "key": api_issue["key"],
            "url": f"{self.base_url}/browse/{api_issue['key']}",
            "fields": {
                "summary": f.get("summary", ""),
                "status": (f.get("status") or {}).get("name"),
                "issuetype": (f.get("issuetype") or {}).get("name"),
                "project": (f.get("project") or {}).get("key"),
                "reporter": (f.get("reporter") or {}).get("displayName"),
                "created": f.get("created"),
                "updated": f.get("updated"),
                "description": f.get("description") or "",
                "comments": comments,
                "acl_groups": [],
            },
        }

File: ts/connectors/jira.ts
/**
 * Live Jira DC connector: cursor-based JQL incremental sync. Output feeds
 * ingest/jira.normalize unchanged. Descriptions are wiki markup — close
 * enough for FTS.
 */

import type { JiraIssue } from "../ingest/jira.js";
import { type DcClient, type Fetcher, getJson, makeClient } from "./auth.js";
import { cqlTs } from "./confluence.js";
import { stableListing } from "./stable-pages.js";

export const PAGE_SIZE = 50;

/** Cheapest read-only "does this token work" call — never depends on any
 *  project/issue existing or being visible to it. Used only by `eil doctor`. */
export async function doctorProbe(client: DcClient): Promise<void> {
  await getJson(client, "/rest/api/2/myself");
}
// issuelinks is the point of this list. EIL was regex-scraping ticket keys out
// of prose while Jira's OWN structured dependency graph — blocks, is-blocked-by,
// duplicates, relates-to — sat one field away and was never requested. assignee,
// labels, resolution and priority are the facets an SDLC agent filters on, and
// parent ties a subtask to its epic.
const FIELDS = [
  "summary",
  "description",
  "status",
  "issuetype",
  "project",
  "reporter",
  "created",
  "updated",
  "comment",
  "assignee",
  "labels",
  "priority",
  "resolution",
  "resolutiondate",
  "components",
  "fixVersions",
  "parent",
  "issuelinks",
].join(",");

export class JiraClient {
  readonly client: DcClient;

  constructor(baseUrl?: string, token?: string, fetcher?: Fetcher) {
    this.client = makeClient("JIRA", baseUrl, token, fetcher);
  }

  async *updatedSince(cursor: string | null, scope?: string): AsyncGenerator<JiraIssue> {
    const clauses: string[] = [];
    if (scope) clauses.push(scope);
    if (cursor) clauses.push(`updated >= "${cqlTs(cursor)}"`);
    const where = clauses.length > 0 ? `${clauses.join(" and ")} ` : "";
    const jql = `${where}order by updated asc, key asc`;
    const issues = await stableListing(
      "Jira incremental listing",
      () => this.scanIssues(jql, "key,updated"),
      (issue) => String(issue.key),
    );
    // Fetch and yield bodies sequentially after the cheap inventory agrees so
    // successful items advance the durable cursor before a later fetch fails.
    for (const issue of issues) yield await this.getIssue(String(issue.key));
  }

  /** Fetch one issue live — the get_doc fresh=true pull-through (flow K5). */
  async getIssue(key: string): Promise<JiraIssue> {
    const data = await getJson(this.client, `/rest/api/2/issue/${encodeURIComponent(key)}`, {
      fields: FIELDS,
    });
    return this.toIssueDict(data);
  }

  /** Complete id listing for reconcile (flow K1 deletions) — keys only, paged. */
  async listIds(scope?: string): Promise<string[]> {
    const where = scope ? `${scope} ` : "";
    const jql = `${where}order by key asc`;
    const issues = await stableListing(
      "Jira reconciliation listing",
      () => this.scanIssues(jql, "key"),
      (issue) => String(issue.key),
    );
    return issues.map((issue) => `jira:issue:${issue.key}`);
  }

  private async scanIssues(jql: string, fields: string): Promise<any[]> {
    const out: any[] = [];
    let start = 0;
    for (;;) {
      const data = await getJson(this.client, "/rest/api/2/search", {
        jql,
        fields,
        maxResults: PAGE_SIZE,
        startAt: start,
      });
      const issues = data.issues ?? [];
      out.push(...issues);
      start += issues.length;
      if (start >= (data.total ?? 0) || issues.length === 0) return out;
    }
  }

  toIssueDict(apiIssue: any): JiraIssue {
    const f = apiIssue.fields;
    const comments = (f.comment?.comments ?? []).map((c: any) => ({
      author: c.author?.displayName ?? "unknown",
      body: c.body ?? "",
      visibility: c.visibility ?? null,
    }));
    return {
      key: apiIssue.key,
      url: `${this.client.baseUrl}/browse/${apiIssue.key}`,
      fields: {
        summary: f.summary ?? "",
        status: f.status?.name ?? null,
        issuetype: f.issuetype?.name ?? null,
        project: f.project?.key ?? null,
        reporter: f.reporter?.displayName ?? null,
        created: f.created ?? null,
        updated: f.updated ?? null,
        description: f.description ?? "",
        comments,
        acl_groups: [],
        assignee: f.assignee?.displayName ?? null,
        labels: f.labels ?? [],
        priority: f.priority?.name ?? null,
        resolution: f.resolution?.name ?? null,
        components: (f.components ?? []).map((c: any) => c.name).filter(Boolean),
        fix_versions: (f.fixVersions ?? []).map((v: any) => v.name).filter(Boolean),
        parent: f.parent?.key ?? null,
        // Jira reports a link from whichever side it was created, so the
        // relevant key is inward OR outward. Taking only one direction would
        // silently halve the dependency graph.
        //
        // Label and key MUST come from the same side. A Jira link type carries
        // BOTH labels ("blocks" / "is blocked by"), so coalescing them
        // independently — `type.outward ?? type.inward` beside
        // `outwardIssue ?? inwardIssue` — reads the outward LABEL off a link
        // that only has an inward ISSUE, and reports "PAY-1 blocks PAY-2" when
        // the truth is the reverse. Harmless while the type was being
        // discarded; a wrong fact the moment it is persisted.
        issue_links: (f.issuelinks ?? [])
          .map((l: any) =>
            l.outwardIssue?.key
              ? { type: l.type?.outward ?? l.type?.name ?? "relates to", key: l.outwardIssue.key }
              : {
                  type: l.type?.inward ?? l.type?.name ?? "relates to",
                  key: l.inwardIssue?.key ?? "",
                },
          )
          .filter((l: any) => l.key),
      },
    };
  }
}
